libs/hwes/eventstoredb/projections/custom/custom.go

```diff
@@ -0,0 +1,135 @@
+package custom
+
+import (
+	"context"
+	"fmt"
+	"github.com/EventStore/EventStore-Client-Go/esdb"
+	zlog "github.com/rs/zerolog/log"
+	"hwes"
+	"strings"
+)
+
+const subscriptionGroupName = "custom"
+
+type eventHandler func(ctx context.Context, evt hwes.Event) (error, esdb.Nack_Action)
+
+type CustomProjection struct {
+	es            *esdb.Client
+	eventHandlers map[string]eventHandler
+}
+
+func NewCustomProjection(esdbClient *esdb.Client) *CustomProjection {
+	return &CustomProjection{es: esdbClient, eventHandlers: make(map[string]eventHandler, 0)}
+}
+
+func (p *CustomProjection) RegisterEventListener(eventType string, eventHandler eventHandler) *CustomProjection {
+	if _, found := p.eventHandlers[eventType]; found {
+		zlog.Error().
+			Str("eventType", eventType).
+			Msg("override existing event handler")
+	}
+
+	p.eventHandlers[eventType] = eventHandler
+	return p
+}
+
+func (p *CustomProjection) handleEvent(ctx context.Context, event hwes.Event) (error, esdb.Nack_Action) {
+	eventHandler, found := p.eventHandlers[event.EventType]
+	if !found {
+		return fmt.Errorf("event type '%s' is invalid", event.EventType), esdb.Nack_Unknown
+	}
+	return eventHandler(ctx, event)
+}
+
+func (p *CustomProjection) Subscribe(ctx context.Context) error {
```

Comment:
do we need to unsubscribe?

Reply:
also add a comment mentioning this method is thread blocking until the context finishes

Reply:
We are closing the stream at https://github.com/helpwave/services/pull/640/files#diff-0cf98f741b76c7e905a897c32947f1036fd56bc1a01debcba8e343b4435a1e79R76. A explicit unsubscribe is not needed. The state of the client is managed by the server.